monai/metrics/panoptic_quality.py

```diff
@@ -0,0 +1,282 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Union
+
+import torch
+
+from monai.metrics.metric import CumulativeIterationMetric
+from monai.metrics.utils import do_metric_reduction, remap_instance_id
+from monai.utils import MetricReduction, optional_import
+
+linear_sum_assignment, _ = optional_import("scipy.optimize", name="linear_sum_assignment")
+
+__all__ = ["PanopticQualityMetric", "compute_panoptic_quality"]
+
+
+class PanopticQualityMetric(CumulativeIterationMetric):
+    """
+    Compute Panoptic Quality between two tensors. If specifying `metric_name` to "SQ" or "RQ",
+    Segmentation Quality (SQ) or Recognition Quality (RQ) will be returned instead.
+
+    Panoptic Quality is a metric used in panoptic segmentation tasks. This task unifies the typically distinct tasks
+    of semantic segmentation (assign a class label to each pixel) and
+    instance segmentation (detect and segment each object instance). Compared with semantic segmentation, panoptic
+    segmentation distinguish different instances that belong to same class.
+    Compared with instance segmentation, panoptic segmentation does not allow overlap and only one semantic label and
+    one instance id can be assigned to each pixel.
+    Please refer to the following paper for more details:
+    https://openaccess.thecvf.com/content_CVPR_2019/papers/Kirillov_Panoptic_Segmentation_CVPR_2019_paper.pdf
+
+    Different from the original panoptic quality, this class refers to the evaluation metric of CoNiC Challenge 2022,
+    a Nuclei instance segmentation and classification task. The referred implementation can be found in:
+    https://github.com/TissueImageAnalytics/CoNIC
+
+    Args:
+        num_classes: number of classes. The number should not count the background.
+        metric_name: output metric. The value can be "pq", "sq" or "rq".
+        reduction: define mode of reduction to the metrics, will only apply reduction on `not-nan` values,
+            available reduction modes: {``"none"``, ``"mean"``, ``"sum"``, ``"mean_batch"``, ``"sum_batch"``,
+            ``"mean_channel"``, ``"sum_channel"``}, default to `self.reduction`. if "none", will not do reduction.
+        match_iou: IOU threshould to determine the pairing between `y_pred` and `y`. Usually,
+            it should >= 0.5, the pairing between instances of `y_pred` and `y` are identical.
+            If set `match_iou` < 0.5, this function uses Munkres assignment to find the
+            maximal amout of unique pairing.
+        smooth_numerator: a small constant added to the numerator to avoid zero.
+
+    """
+
+    def __init__(
+        self,
+        num_classes: int,
+        metric_name: str = "pq",
```

Comment:
In this way if we want all the three outputs (`PQ`, `SQ` and `RQ`), we should run it three times? Would it be possible to return all of them too (or accept the list of str with any combination)?